File: app/src/main/java/id/ac/ui/cs/mobileprogramming/nabilafv/rememberyourlord/ui/activity/AddActViewModel.java
package id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.ui.activity;

import androidx.hilt.lifecycle.ViewModelInject;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.model.Activity;
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.repository.ActivityRepository;
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.utils.State;

public class AddActViewModel extends ViewModel {
    private ActivityRepository activityRepository;
    private MutableLiveData<State<Activity>> activity = new MutableLiveData<State<Activity>>();
    private MutableLiveData<Boolean> isInsert = new MutableLiveData<Boolean>();
    private MutableLiveData<Boolean> isDatePicked = new MutableLiveData<Boolean>();

    @ViewModelInject
    public AddActViewModel(ActivityRepository activityRepository) {
        this.activityRepository = activityRepository;
        activity.setValue(new State<Activity>().init());
        isInsert.setValue(false);
        isDatePicked.setValue(false);
    }

    public LiveData<State<Activity>> getActivity() {
        return activity;
    }

    public LiveData<Boolean> getIsInsert() {
        return isInsert;
    }

    public LiveData<Boolean> getIsDatePicked() {
        return isDatePicked;
    }

    public void setIsDatePicked() {
        isDatePicked.setValue(true);
    }

    public void addActivity(String title, String description, long date) {
        isInsert.setValue(true);
        activity.setValue(new State<Activity>().loading());
        new Thread(new Runnable() {
            public void run() {
                activityRepository.insertActivity(title, description, date);
            }
        }).start();
    }

    public void finishAddingToCalendar() {
        activity.setValue(new State<Activity>().success(null));
    }
}

File: app/src/main/java/id/ac/ui/cs/mobileprogramming/nabilafv/rememberyourlord/ui/activity/AddActViewModel.kt
package id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.ui.activity

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.model.Activity
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.repository.ActivityRepository
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.utils.State
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class AddActViewModel @Inject constructor(private val activityRepository: ActivityRepository) :
    ViewModel() {
    private val activity = MutableLiveData<State<Activity>>()
    private val isInsert = MutableLiveData<Boolean>()
    private val isDatePicked = MutableLiveData<Boolean>()

    fun getActivity(): LiveData<State<Activity>> {
        return activity
    }

    fun getIsDatePicked(): LiveData<Boolean> {
        return isDatePicked
    }

    fun setIsDatePicked() {
        isDatePicked.value = true
    }

    fun addActivity(title: String, description: String, date: Long) {
        isInsert.value = true
        activity.value = State<Activity>().loading()
        viewModelScope.launch(Dispatchers.IO) {
            activityRepository.insertActivity(title, description, date)
        }
    }

    fun finishAddingToCalendar() {
        activity.value = State<Activity>().success(null)
    }

    init {
        activity.value = State<Activity>().init()
        isInsert.value = false
        isDatePicked.value = false
    }
}